zebra-consensus/src/verify.rs

```diff
@@ -0,0 +1,98 @@
+//! Block verification and chain state updates for Zebra.
+//!
+//! Verification occurs in multiple stages:
+//!   - getting blocks (disk- or network-bound)
+//!   - context-free verification of signatures, proofs, and scripts (CPU-bound)
+//!   - context-dependent verification of the chain state (awaits a verified parent block)
+//!
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use std::{
+    error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+
+mod script;
+mod transaction;
+
+/// The trait constraints that we expect from `zebra_state::ZebraState` errors.
+type ZSE = Box<dyn error::Error + Send + Sync + 'static>;
+/// The trait constraints that we expect from the `zebra_state::ZebraState` service.
+/// `ZSF` is the `Future` type for `zebra_state::ZebraState`. This type is generic,
+/// because `tower::Service` function calls require a `Sized` future type.
```

Comment:
if we had .map for services we could box the future to get rid of the generic while still having the type be sized.